src/legacy/core_plugins/kibana/public/discover/np_ready/components/sidebar/discover_sidebar.tsx

```diff
@@ -0,0 +1,300 @@
+import { SavedObject } from '../../../../../../../../core/types';
+import { groupFields } from './lib/group_fields';
+import {
+  IndexPatternFieldList,
+  IndexPatternField,
+  IndexPattern,
+} from '../../../../../../../../plugins/data/public';
+import { AppState } from '../../angular/discover_state';
+import { getDetails } from './lib/get_details';
+import { getDefaultFieldFilter, setFieldFilterProp, filterFieldList } from './lib/field_filter';
+import { getServices } from '../../../kibana_services';
+import { getIndexPatternFieldList } from './lib/get_index_pattern_field_list';
+
+export interface DiscoverSidebarProps {
+  /**
+   * the selected columns displayed in the doc table in discover
+   */
+  columns: string[];
+  /**
+   * a statistics of the distribution of fields in the given hits
+   */
+  fieldCounts: Record<string, number>;
+  /**
+   * hits fetched from ES, displayed in the doc table
+   */
+  hits: Array<Record<string, unknown>>;
+  /**
+   * List of available index patterns
+   */
+  indexPatternList: Array<SavedObject<IndexPatternAttributes>>;
+  /**
+   * Callback function when selecting a field
+   */
+  onAddField: (fieldName: string) => void;
+  /**
+   * Callback function when adding a filter from sidebar
+   */
+  onAddFilter: (field: IndexPatternField | string, value: string, type: '+' | '-') => void;
+  /**
+   * Callback function when removing a field
+   * @param fieldName
+   */
+  onRemoveField: (fieldName: string) => void;
+  /**
+   * Currently selected index pattern
+   */
+  selectedIndexPattern: IndexPattern;
+  /**
+   * Callback function to select another index pattern
+   */
+  setIndexPattern: (id: string) => void;
+  /**
+   * Current app state, used for generating a link to visualize
+   */
+  state: AppState;
+}
+
+export function DiscoverSidebar({
+  columns,
+  fieldCounts,
+  hits,
+  indexPatternList,
+  onAddField,
+  onAddFilter,
+  onRemoveField,
+  selectedIndexPattern,
+  setIndexPattern,
+  state,
+}: DiscoverSidebarProps) {
+  const [openFieldMap, setOpenFieldMap] = useState(new Map());
+  const [showFields, setShowFields] = useState(false);
+  const [fields, setFields] = useState<IndexPatternFieldList | null>(null);
+  const [fieldFilterState, setFieldFilterState] = useState(getDefaultFieldFilter());
+  useEffect(() => {
+    const newFields = getIndexPatternFieldList(selectedIndexPattern, fieldCounts);
+    setFields(newFields);
+  }, [selectedIndexPattern, fieldCounts, hits]);
+
+  if (!selectedIndexPattern || !fields) {
+    return null;
+  }
+
+  const popularLimit = getServices().uiSettings.get('fields:popularLimit');
+  const useShortDots = getServices().uiSettings.get('shortDots:enable');
+  const onChangeFieldSearch = (field: string, value: string | boolean | undefined) => {
+    const newState = setFieldFilterProp(fieldFilterState, field, value);
+    setFieldFilterState(newState);
+  };
+
+  const groupedFields = groupFields(fields, columns, popularLimit, fieldCounts);
```

Comment:
Let's imagine, that the only `showFields` state param will be changed to cause a render.
In that case `groupedFields` will be completely new object which will force to re-render of all `selectedFields`, `popularFields` & `unpopularFields`, which could be skipped since all of them won't be changed

Reply:
thx, i've refactored to use `useMemo`, also added the filtering functionality to this function